server/tso/global_allocator.go

```diff
@@ -84,8 +109,145 @@ func (gta *GlobalTSOAllocator) SetTSO(tso uint64) error {
+	if len(dcLocationMap) == 0 {
+		return gta.timestampOracle.getTS(gta.leadership, count)
+	}
+	// Colletc all allocator leaders' client URLs
+	allocatorLeaders, err := gta.allocatorManager.GetLocalAllocatorLeaders()
+	if err != nil {
+		return pdpb.Timestamp{}, err
+	}
+	leaderURLs := make([]string, 0, len(allocatorLeaders))
+	for _, allocator := range allocatorLeaders {
+		for _, leaderURL := range allocator.GetMember().GetClientUrls() {
+			if slice.NoneOf(leaderURLs, func(i int) bool { return leaderURLs[i] == leaderURL }) {
+				leaderURLs = append(leaderURLs, leaderURL)
+			}
+		}
+	}
+	// Send maxTS to all Local TSO Allocator leaders to prewrite
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	var maxTS time.Time
+	// Collect the MaxTS with all Local TSO Allocator leaders first
+	if maxTS, err = gta.syncMaxTS(ctx, leaderURLs, dcLocationMap, maxTS); err != nil {
+		return pdpb.Timestamp{}, err
+	}
+	maxTS = maxTS.Add(updateTimestampGuard)
+	// Sync the MaxTS with all Local TSO Allocator leaders then
+	if _, err := gta.syncMaxTS(ctx, leaderURLs, dcLocationMap, maxTS); err != nil {
+		return pdpb.Timestamp{}, err
+	}
+	globalTSO := tsoutil.GenerateTimestamp(maxTS, 0)
+	// Update the global TSO in memory
+	if err := gta.SetTSO(tsoutil.GenerateTS(globalTSO)); err != nil {
+		return pdpb.Timestamp{}, err
+	}
+	return *globalTSO, nil
+}
+
+const (
+	dialTimeout = 3 * time.Second
+	rpcTimeout  = 3 * time.Second
+)
+
+func (gta *GlobalTSOAllocator) syncMaxTS(ctx context.Context, leaderURLs []string, dcLocationMap map[string][]uint64, maxTS time.Time) (time.Time, error) {
+	respCh := make(chan *pdpb.SyncMaxTSResponse, len(leaderURLs))
+	errCh := make(chan error, len(leaderURLs))
+	wg := sync.WaitGroup{}
+	for _, leaderURL := range leaderURLs {
+		leaderConn, err := gta.getOrCreateGRPCConn(ctx, leaderURL)
+		if err != nil {
+			return typeutil.ZeroTime, err
+		}
+		wg.Add(1)
+		go func(ctx context.Context, conn *grpc.ClientConn, respCh chan<- *pdpb.SyncMaxTSResponse, errCh chan<- error) {
+			request := &pdpb.SyncMaxTSRequest{
+				Header: &pdpb.RequestHeader{
+					SenderId: gta.allocatorManager.member.ID(),
+				},
+			}
+			if maxTS != typeutil.ZeroTime {
+				request.MaxTs = tsoutil.GenerateTimestamp(maxTS, 0)
+			}
+			syncCtx, cancel := context.WithTimeout(ctx, rpcTimeout)
+			resp, err := pdpb.NewPDClient(conn).SyncMaxTS(syncCtx, request)
+			cancel()
+			if err != nil {
+				errCh <- err
+			}
+			respCh <- resp
+			wg.Done()
+		}(ctx, leaderConn, respCh, errCh)
+	}
+	wg.Wait()
+	close(respCh)
+	close(errCh)
+	// If any error occurs, the synchronization process will fail
+	if err := <-errCh; err != nil {
+		return typeutil.ZeroTime, err
+	}
+	for resp := range respCh {
+		if resp == nil {
+			return typeutil.ZeroTime, errs.ErrSyncMaxTS.FastGenWithCause("got nil response")
+		}
+		if notSyncedDCs := gta.checkSyncedDCs(dcLocationMap, resp.GetDcs()); len(notSyncedDCs) > 0 {
```

Comment:
It looks that if there are multiple responses, it is likely to return error on receiving the first response.

Reply:
And I think it is better to retry instead return an error directly.